baseplate/frameworks/queue_consumer/kombu.py

```diff
@@ -0,0 +1,192 @@
+
+    Raising an Exception that is a subclass of AbortError will cause the
+    KombuMessageHandler to re-raise the exception rather than swallowing it
+    which will cause the handler thread/process to stop.  This, in turn, will
+    gracefully shut down the QueueConsumerServer currently running.
+
+    Exceptions of this nature should be reserved for errors that are due to
+    problems with the environment rather than the message itself.  For example,
+    a node that cannot get its AWS credentials.
+    """
+
+
+class KombuConsumerWorker(ConsumerMixin, PumpWorker):  # pylint: disable=abstract-method
+    """Consumes messages from the given queues and pumps them into the internal work_queue.
+
+    This class does not directly implement the abstract `run` command from
+    PumpWorker because the ConsumerMixin class already defines it.
+    """
+
+    def __init__(
+        self,
+        connection: kombu.Connection,
+        queues: Sequence[kombu.Queue],
+        work_queue: WorkQueue,
+        serializer: Optional[KombuSerializer] = None,
+    ):
+        self.connection = connection
+        self.queues = queues
+        self.work_queue = work_queue
+        self.serializer = serializer
+
+    def get_consumers(self, Consumer: kombu.Consumer, channel: Channel) -> Sequence[kombu.Consumer]:
+        args = dict(queues=self.queues, on_message=self.work_queue.put)
+        if self.serializer:
+            args["accept"] = self.serializer.name
+        return [Consumer(**args)]
+
+    def stop(self) -> None:
+        logger.debug("Closing KombuConsumerWorker.")
+        # `should_stop` is an attribute of `ConsumerMixin`
+        self.should_stop = True
+
+
+class KombuMessageHandler(MessageHandler):
+    def __init__(self, baseplate: Baseplate, name: str, handler_fn: Handler):
+        self.baseplate = baseplate
+        self.name = name
+        self.handler_fn = handler_fn
+
+    def handle(self, message: kombu.Message) -> None:
+        context = self.baseplate.make_context_object()
+        try:
+            # We place the call to ``baseplate.make_server_span`` inside the
+            # try/except block because we still want Baseplate to see and
+            # handle the error (publish it to error reporting)
+            with self.baseplate.make_server_span(context, self.name) as span:
+                delivery_info = message.delivery_info
+                span.set_tag("kind", "consumer")
+                span.set_tag("routing_key", delivery_info.get("routing_key", ""))
+                span.set_tag("consumer_tag", delivery_info.get("consumer_tag", ""))
+                span.set_tag("delivery_tag", delivery_info.get("delivery_tag", ""))
+                span.set_tag("exchange", delivery_info.get("exchange", ""))
+                self.handler_fn(context, message)
+        except Exception as exc:
+            logger.exception(
+                "Unhandled error while trying to process a message, re-queueing the message."
```

Comment:
is there a clearer terminology for requeue that captures the possibility of deadletter queues? (i see the kombu method is called that but this is our messaging to the end user that i want to be clear in)

Reply:
Hmm, maybe say something like `"the message has been re-delivered to the message queue"?  (Redelivered is the term RabbitMQ appears to use).

Reply:
or like "returned to the queue broker" or something?